lib/posix/nanosleep.c

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2018 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <limits.h>
+#include <errno.h>
+/* required for struct timespec */
+#include <posix/time.h>
+
+#define NS_PER_SEC 1000000000ULL
+#define NS_PER_USEC 1000ULL
+
+/**
+ * @brief Suspend execution for nanosecond intervals.
+ *
+ * See IEEE 1003.1
+ */
+int nanosleep(const struct timespec *rqtp, struct timespec *rmtp)
+{
+	u64_t ns;
+	const bool write = (rmtp != NULL);
+
+	if (rqtp == NULL) {
+		errno = EFAULT;
+		return -1;
+	}
+	if (rqtp->tv_sec < 0 || rqtp->tv_nsec < 0
+		|| rqtp->tv_nsec >= NS_PER_SEC) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (rqtp->tv_sec == 0 && rqtp->tv_nsec == 0) {
+		goto maybe_write;
```

Comment:
Label name sounds surprising. It's rather "return_val" or "skip" to me. Probably not worth changing, just thinking aloud.